### ISSUE TITLE
'ids' has some limitations that shouldn't exist

### ISSUE THREAD

**Issue by mkfink** [i3detroit/Automatic-IoT-module-programming#18]
- "commands" can't be included inside "ids"
- the "%id%" template can't be used in parameters inside "ids"
- "ids" is just a bad name
- "id" doesn't need to be defined in all cases (see https://github.com/i3detroit/Automatic-IoT-module-programming/blob/feature/custom-to-tasmota/config/tasmota.json#L502)

**Comment by mtfurlan** [comment]
Why would you want to use %id% in an entry?

**Comment by mkfink** [comment]
Sorry, I was unclear I think. I meant that "%id%" doesn't get replaced with the "id" value if it's inside of the "ids" block.

**Comment by mtfurlan** [comment]
```
    {
       ...
        "ids": [
            {
                "id": "001",
                "topic": "stuff/%id%/otherstuff"
            }
        ]
    }
```
This, yeah? Why would you want to do that?

**Comment by mkfink** [comment]
Oh, I see, you're generally right. The only benefit I can then see is if you wanted to change the id and you've used multiple instances of %id%, then you'd only have to change it in one place:
```
    {
       ...
        "ids": [
            {
                "id": "001",
                "topic": "stuff/%id%/otherstuff",
                "commands": [
                    {
                        "command": "cmnd/%id%/foo",
                        "payload": "bar"
                    }
                ]
            }
        ]
    }
```
But that's pretty minor.

**Comment by mtfurlan** [comment]
I don't see a usecase, but I can do it?
Commands should be appended if the main thing has a command, yeah?

Name proposal: A config is an array of configurations, and a configuration is either a device or an array of devices?

**Comment by mkfink** [comment]
Yeah, commands and probably build flags should be appended.

Names: I think I would put it as a configuration file holds an array of devices and device arrays.

**Comment by mtfurlan** [comment]
I like how reply by email didn't work.
Anyway have the message I sent 17 hours ago.


Gotta have different names for the
```
[
    thingA
    {
        ...thingA
        arrayOfThingB: [
            thingB,
            thingB
        ]
    }
]
```

**Comment by mkfink** [comment]
Ok, how about this. The config file contains an array of device archetypes. Archetypes may or may not contain an array of instances. A device is defined as the union of an archetype and one of its instances. If there are no instances then the device is just defined by the archetype.

![image](https://user-images.githubusercontent.com/11447986/73082871-3b5b1f00-3e98-11ea-8297-56efab9cccb7.png)

**Comment by mtfurlan** [comment]
What about just always requiring an instance?
We don't have many one-off devices, and it's not *that* much overhead.

**Comment by mkfink** [comment]
That's true. I think that's fine. I just skimmed and there are like 5 especially unique devices. All the other individual ones can merge together in some way

**Comment by mtfurlan** [comment]
Alright, so a config is an array of archetypes that have arrays of device instances.

**Comment by mtfurlan** [comment]
Site config?

**Comment by mkfink** [comment]
I took a stab at redoing the config here https://github.com/i3detroit/Automatic-IoT-module-programming/commit/08cd24b35ad22a458a0fa7ed7621a33588c52651, grouping more similar things together and requiring the "instances" field. Doing that, devices got reordered so the diff isn't especially useful to look at. But overall, the file is shorter by a third, but I feel like the change made the file a lot harder for me to skim and understand.

**Comment by mtfurlan** [comment]
Honestly the issues I had reading the single instance ones were pretty easy to overcome in comparison to the lights?
Also having the items in an archetype be in different orders every time is super hard to read.

I have a bad proposal:
```
[
    {
        "name": "Light %id%",
        ...
        instances: [
            {
                "group_topic": "lights/row1"
                 instances: [
                     {
                         "id": "001",
                         "ip_addr": "10.13.107.133"
                     }
                ]
             },
            {
                "group_topic": "lights/row2"
                 instances: [
                     {
                         "id": "00whatever",
                         "ip_addr": "stuff"
                     }
                ]
             },
        ]
    }
]
```

**Comment by mkfink** [comment]
I thought about nesting like that, but I find that to the the opposite of human readable. After dwelling on it a few days, I don't really like requiring 'instances' and would really rather it be used sparsely. I'm leaning a lot more toward just renaming 'ids' to 'instances' and leaving it at that. Repeated info doesn't bother me as much as having to work to understand the file.

**Comment by mtfurlan** [comment]
What do you think about a config level site config?
I keep forgetting to change the site when I copy things to my config.

**Comment by mkfink** [comment]
How do you mean? Something like
mark.json:
```
{
    "site": {
        "mqtt_host": "things",
        "wifi_ssid": "stuff",
        ...
    },
    "devices": [
        {
            "name": "thing1",
            ...
        },
        {
            "name": "thing2",
            ...
        }
    ]
}
```
And then all devices in the file default to use the site that's defined in the file?

**Comment by mtfurlan** [comment]
Yeah, cause there will never be a case where there are multiple sites in one config, file, yeah?
Or keep sites.json, and just have
```
{
    site: "whatever",
    devices: [
        ...
    ]
}
```

**Comment by mkfink** [comment]
Yeah, that wasn't an assumption I originally had, but it seems plausible. I can think of a couple edge cases, like having some devices on one SSID and some on another (if we had something that needed to talk to the internet and we didn't want to set up firewall rules), or running multiple brokers which we don't want to do.

**Comment by mtfurlan** [comment]
So keep sites.json, and if site is in the device it overrides the file?

**Comment by mtfurlan** [comment]
Fixed by #20
